### PR TITLE
(359) Allow reason to be specified when withdrawing application

### PR DESCRIPTION
**Pull request by pezholio**
## Screenshots

image

image

### CODE REVIEW

**Comment by pezholio** [comment]
[Ability to withdraw Placement Requests](https://trello.com/c/jpqfdC7Q/359-ability-to-withdraw-placement-requests)

**Review by richpjames — APPROVED** [review]

**Comment by pezholio** [comment]
[Fix issue with lost bed dates in the calendar not always displaying correctly](https://trello.com/c/R5TSTwce/371-fix-issue-with-lost-bed-dates-in-the-calendar-not-always-displaying-correctly)

File: server/utils/applications/withdrawalReasons.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+const filterByType = <T extends WithdrawalReason>(keys: Array<string>): Record<T, string> => {
+  return Object.keys(withdrawlReasons)
+    .filter(k => keys.includes(k))
+    .reduce((criteria, key) => ({ ...criteria, [key]: withdrawlReasons[key] }), {}) as Record<T, string>
+}
+
+const mapToRadioOptions = <T extends WithdrawalReason>(items: Record<T, string>): Array<RadioItem> =>
+  Object.keys(items).map(k => ({ value: k, text: items[k] }))
```
[the remark this answers is not in the document]
Comment (pezholio) [reply server/utils/applications/withdrawalReasons.ts]: Yeah, we can do that check in the view with `formPageRadios`

File: server/utils/applications/withdrawalReasons.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+export type PlacementNoLongerRequiredReasons = Extract<
+  WithdrawalReason,
+  (typeof placementNoLongerRequiredReasons)[number]
+>
+export type NewApplicationToBeSubmittedReasons = Extract<
+  WithdrawalReason,
+  (typeof newApplicationToBeSubmittedReasons)[number]
+>
+export type ApplicationProblemReasons = Extract<WithdrawalReason, (typeof applicationProblemReasons)[number]>
+export type OtherReasons = Extract<WithdrawalReason, (typeof otherReasons)[number]>
```
[the remark this answers is not in the document]
Comment (pezholio) [reply server/utils/applications/withdrawalReasons.ts:23]: Ah, nice one!

[the remark this answers is not in the document]
Comment (pezholio) [reply server/utils/applications/withdrawalReasons.ts:23]: I've also fixed it in placementCriteriaUtils here https://github.com/ministryofjustice/hmpps-approved-premises-ui/pull/997/commits/7c7ece144c680a9bda58f3235da9aa07c4f5a692

**Pull request merged**